src/EditorFeatures/Core.Wpf/InlineErrors/InlineErrorTag.cs

```diff
@@ -0,0 +1,153 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Windows;
+using System.Windows.Controls;
+using System.Windows.Documents;
+using System.Windows.Media;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editor.Implementation.Adornments;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineErrors
+{
+    internal class InlineErrorTag : GraphicsTag
+    {
+        public const string TagID = "inline error - ";
+        public readonly string ErrorType;
+        private readonly DiagnosticData _diagnostic;
+        public readonly InlineErrorsLocations Location;
+
+        public InlineErrorTag(string errorType, DiagnosticData diagnostic, IEditorFormatMap editorFormatMap, InlineErrorsLocations location)
+            : base(editorFormatMap)
+        {
+            ErrorType = errorType;
+            _diagnostic = diagnostic;
+            Location = location;
+        }
+
+        /// <summary>
+        /// Creates a GraphicsResult object which is the error block based on the geometry and formatting set for the item.
+        /// </summary>
+        public override GraphicsResult GetGraphics(IWpfTextView view, Geometry bounds, TextFormattingRunProperties format)
+        {
+            var block = new TextBlock
+            {
+                FontFamily = format.Typeface.FontFamily,
+                FontSize = 0.75 * format.FontRenderingEmSize,
+                FontStyle = FontStyles.Normal,
+                Foreground = format.ForegroundBrush,
+                Padding = new Thickness(left: 2, top: 0, right: 2, bottom: 0),
+            };
+
+            var id = new Run(_diagnostic.Id);
+            var link = new Hyperlink(id)
+            {
+                NavigateUri = new Uri(_diagnostic.HelpLink)
+            };
+
+            var moniker = GetMoniker();
+            if (moniker is not null)
+            {
+                var image = new CrispImage
+                {
+                    Moniker = (ImageMoniker)moniker,
+                    RenderSize = new Size(16, 16),
+                    Margin = new Thickness(1, 0, 5, 0)
+                };
+
+                var statusImage = new InlineUIContainer(image);
+                block.Inlines.Add(statusImage);
+            }
+
+            link.RequestNavigate += HandleRequestNavigate;
+            block.Inlines.Add(link);
+            block.Inlines.Add(": " + _diagnostic.Message);
```

Comment:
hrmm... possibly ok.  a little concerned about left-to-right languages and how we display things...

Reply:
I assume you meant right-to-left? I can switch it dependent on the language? Do any other features do that?

Reply:
@JoeRobich recently had to check RTL languages in the UI. I'm not sure about the editor...

Reply:
The guidance says if you are composing your UI with in-box WPF controls you should be covered. This concludes my knowledge on this topic.